### PR TITLE
Update `finetune_glue.py`

### PR DESCRIPTION
**Pull request by RK900**

### CODE REVIEW

File: src/training/flava_eval/finetune_glue.py [hunk 1]
```diff
@@ -77,6 +93,24 @@ def parse_args(args):
     parser.add_argument(
         "--seed", type=int, default=0, help="Default random seed."
     )
+    parser.add_argument(
```
Comment (gmittal) [inline src/training/flava_eval/finetune_glue.py:98]: Can we please infer this from tokenizer to avoid bugs?

File: src/training/flava_eval/finetune_glue.py [hunk 2]
```diff
@@ -112,7 +146,7 @@ def step(self, metrics):
 class GLUEDataset(Dataset):
     """GLUE dataset."""
 
-    def __init__(self, task, split, text_key, label_key, tokenizer=None):
+    def __init__(self, task, split, text_key, label_key, separator_token='<end_of_text>', tokenizer=None):
```
Comment (gmittal) [inline src/training/flava_eval/finetune_glue.py]: no defaults please. `separator_token=None`

File: src/training/flava_eval/finetune_glue.py [hunk 3]
```diff
@@ -122,21 +156,33 @@ def __init__(self, task, split, text_key, label_key, tokenizer=None):
 
         self.dataset = load_dataset("glue", task, split=split)
         self.label_key = label_key
-        self.text_key = text_key
+        self.text_key = text_key # list of strings
```
Comment (gmittal) [inline src/training/flava_eval/finetune_glue.py:155]: add `assert` to enforce this expectation

File: src/training/flava_eval/finetune_glue.py [hunk 4]
```diff
@@ -122,21 +156,33 @@ def __init__(self, task, split, text_key, label_key, tokenizer=None):
...
+            text1 = item[self.text_key[0]]
+            text2 = item[self.text_key[1]]
+            label = item[self.label_key]
+            return {
+                'text': self.tokenize([text1 + self.separator_token + text2])[0],
+                'label': label
+            }
+        else:
+            item = self.dataset[idx]
+            text = item[self.text_key[0]]
```
Comment (gmittal) [inline src/training/flava_eval/finetune_glue.py:178]: `assert len(self.text_key) == 1`

File: src/training/flava_eval/finetune_glue.py [hunk 5]
```diff
@@ -165,21 +211,22 @@ def get_task_dataloaders(args):
...
-            text_key="sentence",
+            text_key=TEXT_KEYS[task_name],
             label_key="label",
             tokenizer=tokenizer,
+            separator_token=args.separator_token
         )
         dataloader = DataLoader(
             dataset,
             batch_size=args.batch_size,
             shuffle=True,
```
Comment (gmittal) [inline src/training/flava_eval/finetune_glue.py]: Set an `is_train == (split_name == "train")` variable.

`shuffle=is_train, drop_last=is_train`.

File: src/training/flava_eval/finetune_glue.py [hunk 6]
```diff
@@ -223,9 +269,8 @@ def train_one_epoch(model, data, epoch, optimizer, scheduler, early_stop, device
         label = batch["label"].to(device)
```
Comment (gmittal) [inline src/training/flava_eval/finetune_glue.py:322]: Can you double check that this is correct for the tasks (e.g. no additional postprocessing to predictions or labels is necessary)?

File: src/training/flava_eval/finetune_glue.py [hunk 7]
```diff
@@ -197,10 +244,9 @@ def compute_metrics(model, dataloader, device, args):
             label = batch["label"].to(device)
             samples_seen += text.shape[0]
             logits = model(text)
-            logits = logits.view(-1)
-            label = label.view(-1).float()
-            predictions = torch.sigmoid(logits) > 0.5
-            batch_val_loss = nn.functional.binary_cross_entropy_with_logits(logits, label, reduction='sum')
+            predictions = torch.argmax(logits, dim=-1).float()
```
Comment (gmittal) [inline src/training/flava_eval/finetune_glue.py]: How does the argmax work for the regression task?

File: src/training/flava_eval/finetune_glue.py [hunk 8]
```diff
@@ -223,9 +269,8 @@ def train_one_epoch(model, data, epoch, optimizer, scheduler, early_stop, device
         label = batch["label"].to(device)
 
         logits = model(text)
-        logits = logits.view(-1)
-        label = label.view(-1).float()
-        loss = nn.functional.binary_cross_entropy_with_logits(logits, label)
+        loss_fn = nn.functional.mse_loss if args.task_name == "stsb" else nn.functional.cross_entropy
```
Comment (gmittal) [inline src/training/flava_eval/finetune_glue.py]: this is repeated code, make a method called `get_loss_fn(task_name)` or something like that and call it to get the loss function

File: src/training/flava_eval/finetune_glue.py [hunk 9]
```diff
@@ -122,21 +147,35 @@ def __init__(self, task, split, text_key, label_key, tokenizer=None):
 
         self.dataset = load_dataset("glue", task, split=split)
         self.label_key = label_key
-        self.text_key = text_key
+        self.text_key = text_key # list of strings
+        assert len(self.text_key) <= 2 and isinstance(self.text_key, list)
```
Comment (gmittal) [inline src/training/flava_eval/finetune_glue.py]: should swap the order of the `and`, otherwise it'll crash if it's not an iterable/list

File: src/training/flava_eval/finetune_glue.py [hunk 10]
```diff
@@ -163,29 +202,39 @@ def get_task_metric(task_name):
 def get_task_dataloaders(args):
     tokenizer = get_tokenizer(args.model)
     task_name = args.task_name
+    roberta_tokenizer = isinstance(tokenizer, transformers.models.roberta.tokenization_roberta.RobertaTokenizer) or isinstance(tokenizer, transformers.models.roberta.tokenization_roberta_fast.RobertaTokenizerFast)
```
Comment (gmittal) [inline src/training/flava_eval/finetune_glue.py:210]: I don't think this is quite right since OpenCLIP uses a wrapper class (you can see this in `tokenizer.py`) around the huggingface tokenizers. You can test your implementation by setting `--model= roberta-ViT-B-32` (and leaving `--pretrained` blank)

File: src/training/flava_eval/finetune_glue.py [hunk 11]
```diff
@@ -163,29 +202,39 @@ def get_task_metric(task_name):
...
+    separator_token = '</s>' if roberta_tokenizer else '<end_of_text>'
```
Comment (gmittal) [inline src/training/flava_eval/finetune_glue.py:211]: For HF tokenizers does `tokenizer.special_tokens_map` not work?

File: src/training/flava_eval/finetune_glue.py [hunk 12]
```diff
@@ -163,29 +202,39 @@ def get_task_metric(task_name):
...
-            text_key="sentence",
+            text_key=TEXT_KEYS[task_name],
             label_key="label",
             tokenizer=tokenizer,
+            separator_token=separator_token
         )
         dataloader = DataLoader(
             dataset,
             batch_size=args.batch_size,
             shuffle=True,
```
Comment (gmittal) [inline src/training/flava_eval/finetune_glue.py]: Should be `shuffle=is_train`

File: src/training/flava_eval/finetune_glue.py (hunk 10 above)
Comment (RK900) [reply src/training/flava_eval/finetune_glue.py:210]: yeah i have to do tokenizer.tokenizer

File: src/training/flava_eval/finetune_glue.py (hunk 11 above)
Comment (RK900) [reply src/training/flava_eval/finetune_glue.py:211]: not always, if the tokenizer doesn't have a sep token specified it won't default to the eos token

File: src/training/flava_eval/finetune_glue.py [hunk 13]
```diff
@@ -36,13 +49,16 @@ def parse_args(args):
     parser.add_argument(
         "--epochs", type=int, default=1000, help="Number of epochs to train for."
     )
+    parser.add_argument(
+        "--num-steps", type=int, default=0, help="Number of steps to train for."
```
Comment (gmittal) [inline src/training/flava_eval/finetune_glue.py:55]: why is default 0? if this is the case can you add: 
```python
assert args.num_steps > 0, 'Number of update steps must be greater than 0.'
```

File: src/training/flava_eval/finetune_glue.py [hunk 14]
```diff
@@ -36,13 +49,16 @@ def parse_args(args):
     parser.add_argument(
         "--epochs", type=int, default=1000, help="Number of epochs to train for."
```
Comment (gmittal) [inline src/training/flava_eval/finetune_glue.py:52]: Why do we still have epochs if we've moved to `--num-steps`?

File: src/training/flava_eval/finetune_glue.py [hunk 15]
```diff
@@ -211,6 +264,47 @@ def compute_metrics(model, dataloader, device, args):
...
+    progress_bar = tqdm(total=args.num_steps)
+    data_train = iter(data['train'])
+    for i in range(args.num_steps):
+        scheduler(i)
+
+        try:
+            batch = next(data_train)
+        except StopIteration:
+            data = get_task_dataloaders(args)
+            data_train = iter(data['train'])
```
Comment (gmittal) [inline src/training/flava_eval/finetune_glue.py:280]: Instead of getting the task_dataloaders again can you do the following with the dataloader: `data_iter = itertools.cycle(train_dataloader)` and use that to yield batches.

File: src/training/flava_eval/finetune_glue.py [hunk 16]
```diff
@@ -211,6 +264,47 @@ def compute_metrics(model, dataloader, device, args):
...
+            batch = next(data_train)
+        text = batch["text"].to(device)
+        label = batch["label"].to(device)
+
+        logits = model(text)
+        if args.task_name == "stsb":
+            logits = logits.view(-1)
+            label = label.view(-1).float()
+        loss = get_loss_fn(args.task_name)(logits, label)
```
Comment (gmittal) [inline src/training/flava_eval/finetune_glue.py:289]: you don't need to `get_loss_fn` every training step. outside of the loop you can just do `loss_fn = get_loss_fn(args.task_name)` and then here you can do `loss = lss_fn(logits, labels)`

File: src/training/flava_eval/finetune_glue.py [hunk 17]
```diff
@@ -223,9 +317,10 @@ def train_one_epoch(model, data, epoch, optimizer, scheduler, early_stop, device
         label = batch["label"].to(device)
 
         logits = model(text)
-        logits = logits.view(-1)
-        label = label.view(-1).float()
-        loss = nn.functional.binary_cross_entropy_with_logits(logits, label)
+        if args.task_name == "stsb":
+            logits = logits.view(-1)
+            label = label.view(-1).float()
+        loss = get_loss_fn(args.task_name)(logits, label)
```
Comment (gmittal) [inline src/training/flava_eval/finetune_glue.py:328]: ditto

File: src/training/flava_eval/finetune_glue.py [hunk 18]
```diff
@@ -271,11 +366,16 @@ def main(args):
...
 
-    for epoch in range(args.epochs):
-        val_metrics, end_training = train_one_epoch(clf, data, epoch, optim, scheduler, early_stop, device, args)
+    if args.num_steps > 0:
+        val_metrics, end_training = train_n_steps(clf, data, optim, scheduler, early_stop, device, args)
         if end_training:
             print("Stopped early to prevent overfitting.")
-            break
+    else:
+        for epoch in range(args.epochs):
```
Comment (gmittal) [inline src/training/flava_eval/finetune_glue.py:380]: oh, are you sure we need this still?
